[RedGifsSearch] Add RedGifs search (and user) extractor

Deer-Spangle opened the pull request:
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
  - There is another RedGifs extractor here #1631, but that's pages, rather than searches. But there will be merge conflicts with this due to filename. I'm happy to either rename file, or wait until that one is merged and resolve conflicts!
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

I'm adding a playlist extract for RedGifs searches.
There's another PR Ive noted above, #1631 which does redgifs pages. I'm happy to resolve merge conflicts if that goes in first, and neaten things up. Or to move this to a different file to avoid conflicts. I'm open to your suggestion there.

I wasn't sure whether to implement paging in this also?
As in, should this scrape through all pages, or just the specified page?

MinePlayersPE commented on yt_dlp/extractor/redgifs.py:
```diff
@@ -0,0 +1,98 @@
+from ..utils import ExtractorError, int_or_none
+
+
+def _parse_gif_entry(gif_data):
+    video_id = gif_data["id"]
+    title = " ".join(gif_data.get('tags') or [])
+    formats = [
+        {
+            "format_id": "gif",
+            "url": gif_data["urls"]["gif"]
+        },
+        {
+            "format_id": "sd",
+            "url": gif_data["urls"]["sd"]
+        },
+        {
+            "format_id": "hd",
+            "url": gif_data["urls"]["hd"]
+        }
+    ]
+    webpage_url = "https://redgifs.com/watch/{}".format(video_id)
+
+    return {
+        'id': video_id,
+        'title': title,
+        'age_limit': 18,
+        'formats': formats,
+        'timestamp': gif_data.get('createDate'),
+        'view_count': int_or_none(gif_data.get('views')),
+        'uploader': gif_data.get('userName'),
+        'webpage_url': webpage_url,
+        'tags': gif_data.get('tags')
+    }
```
imo should still be a class function

MinePlayersPE replied:
nvmd, i'll wait for https://github.com/yt-dlp/yt-dlp/pull/1631

Deer-Spangle replied:
It's a class function now, in RedGifsIE, which the other extractors inherit from

MinePlayersPE commented on yt_dlp/extractor/redgifs.py:
```diff
@@ -0,0 +1,98 @@
+
+        tags = query.get("tags")[0]
+        order = query.get("order", ("trending",))[0]
+        page = query.get("page", (1,))[0]
```
maybe try adding an example of this to the tests? it should also just paginate between all pages imo but not one to judge

MinePlayersPE replied:
(was referring to page query)

Deer-Spangle replied:
Yeah, I wasn't sure whether to paginate through all pages, or just do the specified page, to be honest. 
Is there a standard? 
I've added a test of page query in the meanwhile

pukkandan replied:
If there is no page number given in the URL, get all pages. Otherwise, (optionally) you can get just the page in the URL. Note that when paginating, it is best to return a `PagedList` instead of a generator/list (and I think this extractor is able to do that)

Deer-Spangle replied:
Sounds like a great tactic. Thanks for the heads-up on `PagedList`, implemented that now

MinePlayersPE commented on yt_dlp/extractor/redgifs.py:
```diff
@@ -0,0 +1,98 @@
+
+        entries = [_parse_gif_entry(entry) for entry in data['gifs']]
+        title = tags
+        description = "Redgifs search for {}, ordered by {}".format(tags, order)
```
```suggestion
        description = f'Redgifs search for {tags}, ordered by {order}'
```
not a written rule, just something i noticed

Deer-Spangle replied:
Sure, swapped both uses of `string.format()` to f-strings

MinePlayersPE reviewed (changes requested):
p.s. please change all the strings to be single-quoted (except for the one nested inside of the fstring)

pukkandan commented:
#1631 has been merged. Please fix the merge conflicts

Deer-Spangle commented:
Rebased onto master, reworked my extractor to extend RedGifsIE, and pulled out common gif parsing.
I had to narrow the regex of RedGifsIE, to ensure it doesn't match browse pages.
And adding the paging, using the PagedList as you suggested (Thanks!)

pukkandan commented on yt_dlp/extractor/redgifs.py:
```diff
@@ -80,15 +75,110 @@ def _real_extract(self, url):
+            api_query['type'] = query.get('type')[0]
+
+        if query.get('page'):
+            page = query.get('page', (1,))[0]
```
Since its inside if, no default is needed
```suggestion
            page = query.get('page')[0]
```

pukkandan commented on yt_dlp/extractor/redgifs.py:
```diff
@@ -80,15 +75,110 @@ def _real_extract(self, url):
+
+class RedGifsSearchIE(RedGifsIE):
+    IE_DESC = 'Redgifs search'
+    _VALID_URL = r'https?://(?:www\.)?redgifs\.com/browse\?(?P<query>.*)'
```
```suggestion
    _VALID_URL = r'https?://(?:www\.)?redgifs\.com/browse\?(?P<query>[^#]+)'
```

Deer-Spangle replied:
Ah, yep, that's nicer

Deer-Spangle commented:
Also added a RedGifsUser extractor, while I'm here

pukkandan commented on yt_dlp/extractor/redgifs.py:
```diff
@@ -80,15 +75,196 @@ def _real_extract(self, url):
+            'view_count': int_or_none(gif_data.get('views')),
+            'like_count': int_or_none(gif_data.get('likes')),
+            'categories': gif_data.get('tags') or [],
+            'webpage_url': webpage_url,
```
Since one extractor is directly retuning another's result, add `ie_key` and `extractor` as well

Deer-Spangle replied:
I don't see those listed in common.py, unless returning a `'_type': 'url'` for use by another extractor? Since they're all returning full results from a base class, is that needed?

pukkandan replied:
The issue here is that the search extractor is directly returning entries that would normally be extracted by the video extractor. This is why `webpage_url`, `ie_key` and `extractor` has to be set

Deer-Spangle replied:
Ah, fair enough, added those then

pukkandan commented on yt_dlp/extractor/redgifs.py:
```diff
@@ -80,15 +75,196 @@ def _real_extract(self, url):
+        return [self._parse_gif_data(entry) for entry in data['gifs']]
+
+    def _real_extract(self, url):
+        match = self._match_valid_url(url)
+        username = match.group('username')
+        query_str = match.group('query')
```
```suggestion
        username, query_str = self._match_valid_url(url).group('username', 'query')
```

pukkandan commented on yt_dlp/extractor/redgifs.py:
```diff
@@ -80,15 +75,196 @@ def _real_extract(self, url):
+            playlist_id = f'{username}?{query_str}'
+        description = f'RedGifs user {username}'
+
+        if not username:
+            raise ExtractorError('Invalid username', expected=True)
```
unreachable code

Deer-Spangle replied:
Seems like it could be reached if the URL is just `redgifs.com/users/`

Deer-Spangle replied:
Wait, that wouldn't match the regex `+`, oh yeah! My bad, I'll remove that then

pukkandan commented on yt_dlp/extractor/redgifs.py:
```diff
@@ -80,15 +75,196 @@ def _real_extract(self, url):
+
+class RedGifsUserIE(RedGifsIE):
+    IE_DESC = 'Redgifs user'
+    _VALID_URL = r'https?://(?:www\.)?redgifs\.com/users/(?P<username>[^/?#]+)(?:\?(?P<query>.*))?'
```
```suggestion
    _VALID_URL = r'https?://(?:www\.)?redgifs\.com/users/(?P<username>[^/?#]+)(?:\?(?P<query>[^#]+))?'
```

pukkandan commented on yt_dlp/extractor/redgifs.py:
```diff
@@ -80,15 +75,196 @@ def _real_extract(self, url):
+        if not username:
+            raise ExtractorError('Invalid username', expected=True)
+
+        api_query = {
+            'order': 'recent',
+        }
+        page = None
+        if query_str:
+            query = compat_parse_qs(query_str)
+            if query.get('order'):
+                api_query['order'] = query.get('order', ('recent',))[0]
+                description += ", ordered by {order}"
+            if query.get('type'):
+                api_query['type'] = query.get('type')[0]
+            page = query.get('page', (1,))[0]
+
```
lots of redundant code
```suggestion
        query = compat_parse_qs(query_str or '')
        api_query = {
            'order': query.get('order', ('recent',))[0],
        }
        if query.get('type'):
            api_query['type'] = query.get('type')[0]
        page = query.get('page', (None,))[0]
        
        description = f'RedGifs user {username}', ordered by {api_query["order"]}'
```

Deer-Spangle replied:
Ah yes, that's a lot prettier. Changed that, thanks

pukkandan commented on yt_dlp/extractor/redgifs.py:
```diff
@@ -80,15 +75,196 @@ def _real_extract(self, url):
+                api_query['type'] = query.get('type')[0]
+            page = query.get('page', (1,))[0]
+
+        if page is not None:
+            entries = self._fetch_page(playlist_id, username, api_query, page)
+        else:
+            entries = OnDemandPagedList(
+                functools.partial(self._fetch_page, playlist_id, username, api_query),
+                self._PAGE_SIZE)
```
this pattern is repeating in both extractors. Maybe make it a method

Deer-Spangle replied:
Okay, pulled that out into the base extractor

pukkandan commented on yt_dlp/extractor/redgifs.py:
```diff
@@ -80,15 +75,196 @@ def _real_extract(self, url):
+        return self._parse_gif_data(video_info['gif'])
+
+
+class RedGifsSearchIE(RedGifsIE):
```
Do not subclass from another user-facing extractor - that's just asking for trouble. Make a `RedGifsBaseInfoExtractor` from which all of these can be subclassed

Deer-Spangle replied:
Makes sense, was wondering that. Done

Deer-Spangle commented:
Thanks for the cleanup there!

pukkandan commented on yt_dlp/extractor/redgifs.py:
```diff
@@ -1,21 +1,96 @@
+            })
+        self._sort_formats(formats)
+
+        webpage_url = f'https://redgifs.com/watch/{video_id}'
+
+        return {
+            'id': video_id,
+            'webpage_url': webpage_url,
```
```suggestion
        return {
            'id': video_id,
            'webpage_url': f'https://redgifs.com/watch/{video_id}',
```
